src/main/kotlin/org/rust/ide/utils/CallInfo.kt

```diff
@@ -22,27 +30,79 @@ class CallInfo private constructor(
     companion object {
         fun resolve(call: RsCallExpr): CallInfo? {
             val fn = (call.expr as? RsPathExpr)?.path?.reference?.resolve() ?: return null
-            if (fn is RsFunction) return CallInfo(fn)
-
             val ty = call.expr.type
-            if (ty is TyFunction) {
-                val parameterNames = getParameterNames(fn) ?: List(ty.paramTypes.size) { "_" }
-                return CallInfo(ty, parameterNames)
+
+            return when {
+                fn is RsFunction -> buildFunctionParameters(fn, ty)
+                ty is TyFunction -> buildFunctionLike(fn, ty)
+                else -> null
             }
-            return null
         }
 
-        private fun getParameterNames(fn: RsElement): List<String>? {
-            val variant = fn as? RsEnumVariant ?: return null
-            return variant.positionalFields.map { (it.typeReference.type as? TyTypeParameter)?.name ?: "_" }
+        private fun buildFunctionLike(fn: RsElement, ty: TyFunction): CallInfo {
+            val parameters = getFunctionLikeParameters(fn) ?: List(ty.paramTypes.size) { "_" to null }
+            return CallInfo(buildParameters(ty.paramTypes, parameters))
+        }
+
+        private fun getFunctionLikeParameters(element: RsElement): List<Pair<String, RsTypeReference?>>? {
+            return when {
+                element is RsEnumVariant -> element.positionalFields.map { "_" to it.typeReference }
+                element is RsStructItem && element.isTupleStruct ->
+                    element.tupleFields?.tupleFieldDeclList?.map { "_" to it.typeReference }
+                element is RsPatBinding -> {
+                    val decl = element.parentOfType<RsLetDecl>() ?: return null
+                    val lambda = decl.expr as? RsLambdaExpr ?: return null
+                    lambda.valueParameters.map { (it.patText ?: "_") to it.typeReference }
+                }
+                else -> null
+            }
         }
 
         fun resolve(methodCall: RsMethodCall): CallInfo? {
-            return (methodCall.reference.resolve() as? RsFunction)?.let { CallInfo(it) }
+            val function = (methodCall.reference.resolve() as? RsFunction) ?: return null
+            val type = methodCall.inference?.getResolvedMethodType(methodCall) ?: return null
+            return buildFunctionParameters(function, type)
+        }
+
+        private fun buildFunctionParameters(function: RsFunction, ty: Ty): CallInfo {
+            val types = let {
+                val types = (ty as? TyFunction)?.paramTypes
+                    ?: function.valueParameters.mapNotNull { it.typeReference?.type }
```

Comment:
`self` is actually filled in another place (https://github.com/intellij-rust/intellij-rust/pull/6664/files#diff-59b6929890a0d4ca48aad7ebe026cff696f8b855e7a3ef42076b819648b1caf8R107), so it works (there are several tests that work with `self`, even with UFCS).

The parameters handle this kind of automatically, since we access `valueParameters`, which doesn't contain `self`. But `TyFunction.paramTypes` does (check e.g. `RsParameterInfoHandlerTest::test method`), so we need to drop the self type from it. It wasn't needed before this PR, because the actual parameter types were not used, just `valueParameters` were used before.

But what is wrong is that we mix `paramTypes` with `valueParameters` here and then drop the first one, but if `valueParameters` would be used, that would indeed result in a skipped parameter. Since `ty` is always `TyFunction`, we can solve that by just checking that `ty` is `TyFunction` soon (in `resolve` and then this problem should go away). I pushed the corresponding change.